Add spec.md to the repository (#315)
ballerina-platform/module-ballerinax-sfdc: docs/spec/spec.md

It offers intuitive resource methods to interact with the Salesforce API v59.

## 2. [Client](#2-client)

`salesforce:Client` can be used to access the Salesforce REST API. 

### 2.1. [Client Configurations](#21-client-configurations)

When initializing the client, following configurations can be provided,

```ballerina
public type ConnectionConfig record {|
    *config:ConnectionConfig;
    # The Salesforce endpoint URL
    string baseUrl;
    # Configurations related to client authentication
    http:BearerTokenConfig|
        config:OAuth2RefreshTokenGrantConfig | config:OAuth2PasswordGrantConfig | config:OAuth2ClientCredentialsGrantConfig auth;
    |}
```

### 2.2. [Initialization](#22-initialization)

A client can be initialized by providing the Salesforce and optionally the other configurations in `ClientConfiguration`.

```ballerina
ConnectionConfig config = {
    baseUrl: salesforceBaseUrl,
    auth: {
        token: salesforceAccessToken
    }
};

Client salesforceClient = check new (config);
```

### 2.3 [APIs](#23-apis)

#### [getOrganizationMetaData](#getOrganizationMetaData)

Used to describe SObjects in the Salesforce

``` ballerina

> //Describe SObjects

Reviewer: Do we need this line?

#
# + sObjectName - sObject name value
# + return - `SObjectMetaData` record if successful or else`error`
isolated remote function describe(string sObjectName) returns SObjectMetaData|error
```

#### [getPlatformAction](#getPlatformAction)

Used to query for actions displayed in the UI, given a user, a context, device format, and a record ID.

``` ballerina
# Query for actions displayed in the UI, given a user, a context, device format, and a record ID.
#
# + return - `SObjectBasicInfo` record if successful or else `error`
isolated remote function getPlatformAction() returns SObjectBasicInfo|error
```

#### [getApiVersions](#getApiVersions)

Used to list summary details about each REST API version available.

``` ballerina
# Lists summary details about each REST API version available.
#
# + return - List of `Version` if successful. Else, the occured `error`.
isolated remote function getApiVersions() returns Version[]|error
```

#### [getResources](#getResources)

Used to list the resources available for the specified API version.

``` ballerina
# Lists the resources available for the specified API version.
#
# + apiVersion - API version (v37)

> # + return - `Resources` as map of strings if successful. Else, the occurred `error`.

Reviewer: ```suggestion
# + return - `Resources` as a map of string if successful. Else, the occurred `error`.
```
WDYT?

isolated remote function getResources(string apiVersion) returns map<string>|error 
```

#### [getLimits](#getLimits)

Used to list the Limits information for your organization.

``` ballerina
# Lists the Limits information for your organization.
#

> # + return - `OrganizationLimits` as map of `Limit` if successful. Else, the occurred `error`.

Reviewer: ```suggestion
# + return - `OrganizationLimits` as a map of `Limit` if successful. Else, the occurred `error`.
```

# + id - sObject ID
# + returnType - The payload, which is expected to be returned after data binding.
# + return - Record if successful or else `error`
isolated remote function getById(string sobjectName, string id, typedesc<record {}> returnType = <>)
                                    returns returnType|error
```
    
#### [getByExternalId](#getByExternalId)

Used to get an object record by external ID.

``` ballerina
# Gets an object record by external ID.
#
# + sobjectName - sObject name 
# + extIdField - External ID field name 
# + extId - External ID value 
# + returnType - The payload, which is expected to be returned after data binding.
# + return - Record if successful or else `error`
isolated remote function getByExternalId(string sobjectName, string extIdField, string extId,
            typedesc<record {}> returnType = <>) returns returnType|error
```

#### [create](#create)

Used to create records based on relevant object type sent with json record.

``` ballerina
# Creates records based on relevant object type sent with json record.
#
# + sObjectName - sObject name value
# + sObject - Record to be inserted
# + return - Creation response if successful or else `error`

> 

Reviewer: ```suggestion
```

isolated remote function create(string sObjectName, record {} sObject)
                                    returns CreationResponse|error 
```

#### [update](#update)

Used to update records based on relevant object ID.

``` ballerina
# Updates records based on relevant object ID.
#
# + sObjectName - sObject name value
# + id - sObject ID
# + sObject - Record to be updated

> # + return - Empty response if successful `error`

Reviewer: ```suggestion
# + return - Nil on success, else returns an error
```

Reviewer: Check other places as well

isolated remote function update(string sObjectName, string id, record {} sObject)
                                    returns error?
```

#### [upsert](#upsert)

Used to upsert a record based on the value of a specified external ID field.

``` ballerina
# Upsert a record based on the value of a specified external ID field.
#
# + sObjectName - sObject name value
# + externalIdField - External ID field of an object
# + externalId - External ID
# + sObject - Record to be upserted
# + return - Empty response if successful or else `error`

isolated remote function upsert(string sObjectName, string externalIdField, string externalId,
            record {} sObject) returns error?
```

#### [delete](#delete)

Used to delete existing records based on relevant object ID.

``` ballerina
# Delete existing records based on relevant object ID.
#
# + sObjectName - SObject name value
# + id - SObject ID
# + return - Empty response if successful or else `error`

> isolated remote function delete( string sObjectName, string id) returns error? 

Reviewer: ```suggestion
isolated remote function delete(string sObjectName, string id) returns error? 
```

```

#### [listReports](#listReports)

Used to list reports

``` ballerina
# Lists reports.
#

> # + return - Array of Report if successful or else `error`

Reviewer: ```suggestion
# + return - Array of `Report` if successful or else `error`
```

isolated remote function listReports() returns Report[]|error
```

> #### [listReports](#listReports)

Reviewer: Same API as the above is given here. Should be removed?

Used to delete existing records based on relevant object ID.

``` ballerina
# Lists reports.
#
# + return - Array of Report if successful or else `error`

isolated remote function listReports() returns Report[]|error
```

#### [deleteReport](#deleteReport)

Used to delete existing report based on relevant report ID.

``` ballerina
# Deletes a report.
#
# + reportId - Report Id

> # + return - `()` if the report deletion is successful or else an error

Reviewer: ```suggestion
# + return - `Nil` if the report deletion is successful or else an error
```

isolated remote function deleteReport(string reportId) returns error?
```

#### [runReportSync](#runReportSync)

Used to run an instance of a report synchronously.

``` ballerina
# Runs an instance of a report synchronously.
#
# + reportId - Report Id
# + return - ReportInstanceResult if successful or else `error`

> @display {label: "Run Report Synchronously"}

Reviewer: @display {label: "Run Report Synchronously"}
Do we need this line?

#
# + sObjectName - SObject reference
# + layoutName - Name of the layout.
# + returnType - The payload type, which is expected to be returned after data binding
# + return - record of `returnType` if successful or else `error`
isolated remote function getNamedLayouts(@display {label: "Name of the sObject"} string sObjectName,
            @display {label: "Name of the layout"} string layoutName, typedesc<record {}> returnType = <>)
                                    returns returnType|error
```

#### [getInvocableActions](#getInvocableActions)

Used to retrieve a list of general action types for the current organization.

``` ballerina
# Retrieve a list of general action types for the current organization.
#
# + subContext - Sub context
# + returnType - The payload type, which is expected to be returned after data binding
# + return - record of `returnType` if successful or else `error`
isolated remote function getInvocableActions(string subContext,
            typedesc<record {}> returnType = <>) returns returnType|error
```

#### [invokeActions](#invokeActions)

Used to invoke actions.

``` ballerina
# Invoke Actions.
#
# + subContext - Sub context

> # + payload - payload for the action

Reviewer: ```suggestion
# + payload - Payload for the action
```

# + returnType - The type of the returned variable
# + return - record of `returnType` if successful or else `error`
isolated remote function invokeActions(string subContext, record {} payload,
            typedesc<record {}> returnType = <>) returns returnType|error
```

#### [deleteRecordsUsingExtId](#deleteRecordsUsingExtId)

Used to delete record using external Id.

``` ballerina
# Delete record using external Id.
#
# + sObjectName - Name of the sObject
# + externalId - Name of the external id field
# + value - value of the external id field
# + return - `()` if successful or else `error`
isolated remote function deleteRecordsUsingExtId(string sObjectName, string externalId, string value) returns error?
```

> #### [deleteRecordsUsingExtId](#deleteRecordsUsingExtId)

Reviewer: ```suggestion
#### [apexRestExecute](#apexRestExecute)
```
API given is different from the description

# + return - `BulkJob` if successful or else `error`
isolated remote function createQueryJob(BulkCreatePayload payload) returns BulkJob|error
```

#### [createQueryJobAndWait](#createQueryJobAndWait)

Used to create a bulkv2 query job and provide a future value.

``` ballerina
# Creates a bulkv2 query job and provide a future value.
#
# + payload - The payload for the bulk job
# + return - `future<BulkJobInfo>` if successful else `error`
isolated remote function createQueryJobAndWait(BulkCreatePayload payload) returns future<BulkJobInfo|error>|error
```

#### [getJobInfo](#getJobInfo)

Used to retrieve detailed information about a job.

``` ballerina
# Retrieves detailed information about a job.
#
# + bulkJobId - Id of the bulk job.
# + bulkOperation - The processing operation for the job.
# + return - `BulkJobInfo` if successful or else `error`
isolated remote function getJobInfo(string bulkJobId, BulkOperation bulkOperation) returns BulkJobInfo|error
```

#### [addBatch](#addBatch)

Used to upload data for a job using CSV data.

``` ballerina
# Uploads data for a job using CSV data.
#
# + bulkJobId - Id of the bulk job
# + content - CSV data to be added

> # + return - `BulkJobInfo` record if successful or `error` if unsuccessful

Reviewer: The return type is `error?`

isolated remote function addBatch(string bulkJobId, string|string[][]|stream<string[], error?>|io:ReadableByteChannel content) returns error?
```

#### [getAllJobs](#getAllJobs)

Used to get details of all the jobs.

``` ballerina
# Get details of all the jobs.
#
# + jobType - Type of the job
# + return - `AllJobs` record if successful or `error` if unsuccessful

> isolated remote function getAllJobs(JobType? jobType = ()) returns error|AllJobs

Reviewer: Shall we change the API to 
```suggestion
isolated remote function getAllJobs(JobType? jobType = ()) returns AllJobs|error
```

Reviewer: Please check other places as well.
